routers/web/repo/pull_review.go

```diff
@@ -242,3 +245,50 @@ func DismissReview(ctx *context.Context) {
 
 	ctx.Redirect(fmt.Sprintf("%s/pulls/%d#%s", ctx.Repo.RepoLink, comm.Issue.Index, comm.HashTag()))
 }
+
+const headCommitKey = "_headCommitSHA"
+
+func UpdateViewedFiles(ctx *context.Context) {
+	issue := checkPullInfo(ctx)
+	if ctx.Written() {
+		return
+	}
+	pull := issue.PullRequest
+	updatedFiles := make(map[string]pulls.ViewedState, len(ctx.Req.Form))
+	headCommitSHA := ""
+
+	// Collect all files and their viewed state
+	for file, values := range ctx.Req.Form {
+		for _, viewedString := range values {
+			viewed, err := strconv.ParseBool(viewedString)
+			// Ignore fields that do not parse as a boolean, i.e. the CSRF token
+			if err != nil {
+
+				// Prevent invalid reviews by specifically supplying the commit the user viewed the file under
+				if file == headCommitKey {
+					headCommitSHA = viewedString
+				}
+				continue
+			}
+
+			// Only unviewed and viewed are possible, has-changed can not be set from the outside
+			state := pulls.Unviewed
+			if viewed {
+				state = pulls.Viewed
+			}
+			updatedFiles[strings.ReplaceAll(file, "%22", "\"")] = state
```

Comment:
:) Why we need a replace here? I think all form value should be escaped?

Reply:
> // \" is the only character that gets encoded when sent as form-encoded string.
			// Unfortunately, this WILL break marking any file as viewed that contains an actual "%22" in its name.
			// There is no way to prevent this, and "%22" is way more unlikely to occur in a filename than \".

Tested on both Chromium and Firefox. Only with this replace, files with a `"` will be parsed correctly.

Reply:
This design is not ideal. 

You had better POST a json to backend, for example: `{"headCommitSHA": ", "reviewdFiles": {....}}`

Do not do `if file == headCommitKey { ... }`

Reply:
How do I POST a JSON?
`fetch()`?

Reply:
There are many examples about how to use `fetch` online. jQuery.ajax/post could also do the work, however I do not know why some people dislike it.

A little off topic: to make the frontend more clear, there could be a general ajax function for Gitea, for example, such function should do the json stringify and add csrf-token automatically, and maybe show errors automatically , then it would reduce a lot of duplicate code. A healthy code base should always be refactoring with pretty functions, instead of copying csrf-token header again and again.